Kernel/Net/NetworkTask.cpp

```diff
@@ -27,59 +28,53 @@ static void handle_icmp(const EthernetFrameHeader&, const IPv4Packet&);
 static void handle_udp(const IPv4Packet&);
 static void handle_tcp(const IPv4Packet&);
 
-Lockable<HashMap<IPv4Address, MACAddress>>& arp_table()
-{
-    static Lockable<HashMap<IPv4Address, MACAddress>>* the;
-    if (!the)
-        the = new Lockable<HashMap<IPv4Address, MACAddress>>;
-    return *the;
-}
-
 void NetworkTask_main()
 {
-    LoopbackAdapter::the();
-
-    auto e1000 = E1000NetworkAdapter::the();
-    if (!e1000)
-        dbgprintf("E1000 network card not found!\n");
-    if (e1000)
-        e1000->set_ipv4_address(IPv4Address(192, 168, 5, 2));
-
-    auto rtl8139 = RTL8139NetworkAdapter::the();
-    if (!rtl8139)
-        dbgprintf("RTL8139 network card not found!\n");
-    if (rtl8139)
-        rtl8139->set_ipv4_address(IPv4Address(192, 168, 13, 201));
-
-    auto dequeue_packet = [&]() -> Optional<KBuffer> {
-        auto packet = LoopbackAdapter::the().dequeue_packet();
-        if (packet.has_value()) {
-            dbgprintf("Receive loopback packet (%d bytes)\n", packet.value().size());
-            return packet.value();
+    u8 octet = 15;
+    int pending_packets = 0;
+    NetworkAdapter::for_each([&octet, &pending_packets](auto& adapter) {
+        if (String(adapter.class_name()) == "LoopbackAdapter") {
+            adapter.set_ipv4_address({ 127, 0, 0, 1 });
+            adapter.set_ipv4_netmask({ 255, 0, 0, 0 });
+            adapter.set_ipv4_gateway({ 0, 0, 0, 0 });
+        } else {
+            adapter.set_ipv4_address({ 10, 0, 2, octet++ });
+            adapter.set_ipv4_netmask({ 255, 255, 255, 0 });
+            adapter.set_ipv4_gateway({ 10, 0, 2, 2 });
         }
-        if (e1000 && e1000->has_queued_packets())
-            return e1000->dequeue_packet();
-        if (rtl8139 && rtl8139->has_queued_packets())
-            return rtl8139->dequeue_packet();
-        return {};
+
+        kprintf("NetworkTask: %s network adapter found: hw=%s address=%s netmask=%s gateway=%s\n",
+            adapter.class_name(),
+            adapter.mac_address().to_string().characters(),
+            adapter.ipv4_address().to_string().characters(),
+            adapter.ipv4_netmask().to_string().characters(),
+            adapter.ipv4_gateway().to_string().characters());
+
+        adapter.on_receive = [&pending_packets]() {
+            pending_packets++;
```

Comment:
This feels like it would be a race condition in a multiple-CPU world - should I do some kind of atomic CAS/increment thing here?